### PR TITLE
Tmt18plex

### LINKED ISSUE #5693: TMTpro 18plex
Hi, I have been warned that TMTpro 18plex is coming (to our lab), and I wanted to ask if there is planned or existing support for it in OpenMS, as I could not find a mention of it in github issues. Thank you :)

https://www.thermofisher.com/order/catalog/product/A52045
https://pubmed.ncbi.nlm.nih.gov/33900084/

### PR DESCRIPTION
**Pull request by jpfeuffer**
## Description

fixes #5693

Adds support for TMT18plex and unlimited affected channels.

In theory, we could leave out "-1" affected channels now and equally reduce the entries in the correction matrix but I kept it
to be on par with the product sheets for now.

Some tests are failing because the old specification was most likely wrong. See my comment https://github.com/OpenMS/OpenMS/pull/6131#issuecomment-1284026575

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by jpfeuffer** [comment]
supersedes #6131

**Comment by jpfeuffer** [comment]
one could argue that isotopes that do not overlap with another channel (e.g. at the end of the channels of TMT16) should be deducted from self-contribution, so I will implement the old way.

File: src/openms/source/ANALYSIS/QUANTITATION/TMTSixteenPlexQuantitationMethod.cpp [hunk 1]
```diff
@@ -110,26 +93,30 @@ void TMTSixteenPlexQuantitationMethod::setDefaultParams_()
...
+                                              "0.08/NA/3.90/NA/0.47/3.57/0.00/0.04",
+                                              "0.15/0.01/3.58/0.72/NA/1.80/NA/0.00",
+                                              "0.11/NA/4.55/NA/0.43/1.86/0.00/0.00",
+                                              "0.07/0.01/3.14/0.73/NA/3.40/NA/0.03",
+                                              "0.22/NA/4.96/NA/0.34/1.03/0.00/NA",
+                                              "0.30/0.03/5.49/0.62/NA/1.14/NA/NA"
+                                            },
+                       "Correction matrix for isotope distributions in percent (see documentation);"
+                       " use the following format: <-2C13>/<-N15-C13>/<-C13>/<-N15>/<+N15>/<+C13>/<+N15+C13>/<+2C13> for each entry of"
+                       " a comma separated list; e.g., 'NA/0.00/0.82/0.65/NA/8.13/NA/0.26, 0.13/NA/2.59/NA/0.32/6.07/0.1/0.09'");
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/QUANTITATION/TMTSixteenPlexQuantitationMethod.cpp:119]:
```suggestion
                       "Correction matrix for isotope distributions in percent from the Thermo data sheet (see documentation);"
                       " Please provide 16 entries (rows), separated by comma, where each entry contains 8 values in the following format: <-2C13>/<-N15-C13>/<-C13>/<-N15>/<+N15>/<+C13>/<+N15+C13>/<+2C13> e.g. one row may look like this: 'NA/0.00  /  0.82/0.65  /  NA/8.13  /  NA/0.26'. You may use whitespaces at your leisure to ease reading.");
```

File: src/openms/source/ANALYSIS/QUANTITATION/TMTEighteenPlexQuantitationMethod.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                                              "0.11/NA/4.55/NA/0.43/1.86/0.00/0.00",
+                                              "0.07/0.01/3.14/0.73/NA/3.40/NA/0.03",
+                                              "0.22/NA/4.96/NA/0.34/1.03/0.00/NA",
+                                              "0.30/0.03/5.49/0.62/NA/1.14/NA/NA",
+                                              "0.14/NA/5.81/NA/0.31/NA/NA/NA",
+                                              "0.19/0.02/5.42/0.36/NA/NA/NA/NA"
+                                            },
+                       "Correction matrix for isotope distributions in percent (see documentation);"
+                       " use the following format: <-2C13>/<-N15-C13>/<-C13>/<-N15>/<+N15>/<+C13>/<+N15+C13>/<+2C13> for each entry of"
+                       " a comma separated list; e.g., 'NA/0.00/0.82/0.65/NA/8.13/NA/0.26, 0.13/NA/2.59/NA/0.32/6.07/0.1/0.09'");
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/QUANTITATION/TMTEighteenPlexQuantitationMethod.cpp:126]:
```suggestion
                       "Correction matrix for isotope distributions in percent from the Thermo data sheet (see documentation);"
                       " Please provide 18 entries (rows), separated by comma, where each entry contains 8 values in the following format: <-2C13>/<-N15-C13>/<-C13>/<-N15>/<+N15>/<+C13>/<+N15+C13>/<+2C13> e.g. one row may look like this: 'NA/0.00  /  0.82/0.65  /  NA/8.13  /  NA/0.26'. You may use whitespaces at your leisure to ease reading.");
```

File: src/openms/source/ANALYSIS/QUANTITATION/TMTSixteenPlexQuantitationMethod.cpp [hunk 3]
```diff
@@ -110,26 +93,30 @@ void TMTSixteenPlexQuantitationMethod::setDefaultParams_()
...
+                                              "0.13/NA/2.59/NA/0.32/6.07/0.1/0.09",
+                                              "0.13/0.00/2.41/0.27/NA/5.58/NA/0.10",
+                                              "0.04/NA/3.10/NA/0.42/4.82/0.02/0.06",
+                                              "0.03/0.00/2.78/0.63/NA/4.57/NA/0.12",
+                                              "0.08/NA/3.90/NA/0.47/3.57/0.00/0.04",
+                                              "0.15/0.01/3.58/0.72/NA/1.80/NA/0.00",
+                                              "0.11/NA/4.55/NA/0.43/1.86/0.00/0.00",
+                                              "0.07/0.01/3.14/0.73/NA/3.40/NA/0.03",
+                                              "0.22/NA/4.96/NA/0.34/1.03/0.00/NA",
+                                              "0.30/0.03/5.49/0.62/NA/1.14/NA/NA"
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/QUANTITATION/TMTSixteenPlexQuantitationMethod.cpp:115]: may we can assist the user with a bit of preformatting here?
```suggestion
                                              "NA/NA / NA/NA / 0.31/9.09 / 0.02/0.32",
                                              "NA/NA / NA/0.78 / NA/9.41 / NA/0.33",
                                              "NA/NA / 0.93/NA / 0.35/8.63 / 0.01/0.27",
                                              "NA/0.00 / 0.82/0.65 / NA/8.13 / NA/0.26",
                                              "0.00/NA / 1.47/NA / 0.34/6.91 / 0.00/0.15",
                                              "0.00/0.00 / 1.46/1.28 / NA/6.86 / NA/0.15",
                                              "0.13/NA / 2.59/NA / 0.32/6.07 / 0.1/0.09",
                                              "0.13/0.00 / 2.41/0.27 / NA/5.58 / NA/0.10",
                                              "0.04/NA / 3.10/NA / 0.42/4.82 / 0.02/0.06",
                                              "0.03/0.00 / 2.78/0.63 / NA/4.57 / NA/0.12",
                                              "0.08/NA / 3.90/NA / 0.47/3.57 / 0.00/0.04",
                                              "0.15/0.01 / 3.58/0.72 / NA/1.80 / NA/0.00",
                                              "0.11/NA / 4.55/NA / 0.43/1.86 / 0.00/0.00",
                                              "0.07/0.01 / 3.14/0.73 / NA/3.40 / NA/0.03",
                                              "0.22/NA / 4.96/NA / 0.34/1.03 / 0.00/NA",
                                              "0.30/0.03 / 5.49/0.62 / NA/1.14 / NA/NA"
```

Comment (cbielow) [reply src/openms/source/ANALYSIS/QUANTITATION/TMTSixteenPlexQuantitationMethod.cpp:115]: same for 18plex (if you agree) :)

File: src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp [hunk 4]
```diff
@@ -40,89 +40,57 @@
...
-    for (StringList::const_iterator it = stringlist.begin(); it != stringlist.end(); ++it)
+    for (const auto& l : stringlist)
     {
       StringList corrections;
-      it->split('/', corrections);
-      if (corrections.size() != 4)
-      {
-        throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "IsobaricQuantitationMethod: Invalid entry in string representation of the isotope correction matrx. Expected four correction values separated by '/', got: '" + *it + "'");
-      }
+      l.split('/', corrections);
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp:68]: Add a check, just to make sure that the user has entered 4 or 8 numbers per line (and not 9 or so...).

Warning: blindly coded :)

```suggestion
      l.split('/', corrections);
      auto number_of_columns = getChannelInformation()[contributing_channel].affected_channels.size();
      if (corrections.size() != number_of_columns )
      {
        throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTYFUNCTION, String("Entry #") + contributing_channel + " must contain " + number_of_columns + " values, but has " + corrections.size() + "!");
      }
```

File: src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h [hunk 5]
```diff
@@ -70,35 +71,35 @@ namespace OpenMS
...
-      /// Id of the -2 isotopic channel (== -1 -> no channel)
-      Int channel_id_minus_2;
-      /// Id of the -1 isotopic channel (== -1 -> no channel)
-      Int channel_id_minus_1;
-      /// Id of the +1 isotopic channel (== -1 -> no channel)
-      Int channel_id_plus_1;
-      // Id of the +2 isotopic channel (== -1 -> no channel)
-      Int channel_id_plus_2;
+      IsobaricChannelInformation(String local_name,
+                                 const Int local_id,
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:92]: ditch the C'tors please and rely on aggregate inititialization (i.e. copy).
C'tors are called a handful of times in the whole program. Not worth just to save a few move's IMHO.

File: src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h [hunk 6]
```diff
@@ -70,35 +71,35 @@ namespace OpenMS
       String description;
       /// The expected centroid position of the channel peak in m/z.
       Peak2D::CoordinateType center;
+      /// Ids of the affected channels. Order has to match the ones from the correction matrix parameter.
+      std::vector<Int> affected_channels;
```
Comment (cbielow) [inline src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:75]:
```suggestion
      /// Ids of the affected channels. Must contain 4 or 8 entries, depending on the number of columns in the Thermo data sheet (with or without subchannels). Order has to match the ones from the correction matrix parameter.
      std::vector<Int> affected_channels;
```

File: src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp (hunk 4 above)
Comment (cbielow) [reply src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp:68]: Could you please add a test, that this actually works? Thanks :)

**Review by cbielow — CHANGES_REQUESTED** [review]
Really really nice!
Thanks!!!

Just a few minor things, mainly formatting of the INI parameter and some consistency checks which would be nice to have.

File: src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:92]: I agree, I am not sure about pyopenms though ;)
Can we just keep it?

File: src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp (hunk 4 above)
Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp:68]: done. almost correct.

File: src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp [hunk 7]
```diff
@@ -40,89 +40,57 @@
...
-        // hence we check if the target_channel corresponds to the contributing_channel -2/-1/+1/+2
-        // if yes, we assign the corresponding contribution
-        // contribution to itself is handled separately
-
-        if (getChannelInformation()[contributing_channel].channel_id_minus_2 == static_cast<Int>(target_channel))
-        {
-          channel_frequency.setValue(target_channel, contributing_channel, isotope_correction_matrix.getValue(contributing_channel, 0) / 100);
-        }
-        else if (getChannelInformation()[contributing_channel].channel_id_minus_1 == static_cast<Int>(target_channel))
+        if (c != "NA" && c != "-1" && c != "0.0")
```
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.cpp:84]: done

File: src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:92]: we have two constructors. One for copy, one for move. I don't see the problem.

**Review by timosachsenberg — APPROVED** [review]
nice PR!

File: src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:92]: It's four lines, 10 mins less work and one cycle of builds less for added flexibility.

Comment (timosachsenberg) [reply src/openms/include/OpenMS/ANALYSIS/QUANTITATION/IsobaricQuantitationMethod.h:92]: If the PR is of high quality, you need to live with the nitpicking ;)
fine as it is

**Review by cbielow — APPROVED** [review]
nice PR.
There are some things which could be overhauled (since they predate our C11 time), but that is for another PR.

**Pull request merged**